### ISSUE TITLE
kokkos build failure on zen3 target 

### ISSUE THREAD

**Issue by shahzebsiddiqui** [LLNL/spack#26950]
### Steps to reproduce

I am doing a `spack install` on Perlmutter using e4s-21.08 branch 

We need to add an entry for zen3 see https://github.com/spack/spack/blob/e4s-21.08/var/spack/repos/builtin/packages/kokkos/package.py#L85-L136 codebase. 

### Error message

```
==> Installing kokkos-3.4.00-a67ukykjrw7m7ixizm5jyaf5d5slvlzk
==> No binary for kokkos-3.4.00-a67ukykjrw7m7ixizm5jyaf5d5slvlzk found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/2e/2e4438f9e4767442d8a55e65d000cc9cde92277d415ab4913a96cd3ad901d317.tar.gz
==> No patches needed for kokkos
==> kokkos: Executing phase: 'cmake'
==> Error: KeyError: 'zen3'

/global/common/software/spackecp/perlmutter/e4s-21.08/spack/var/spack/repos/builtin/packages/kokkos/package.py:253, in cmake_args:
        250                    kokkos_arch_name = self.spack_cuda_arch_map[cuda_arch]
        251                    spack_microarches.append(kokkos_arch_name)
        252
  >>    253        kokkos_microarch_name = self.spack_micro_arch_map[spec.target.name]
        254        if kokkos_microarch_name:
        255            spack_microarches.append(kokkos_microarch_name)
        256

See build log for details:
  /tmp/spack-stage/siddiq90/spack-stage-kokkos-3.4.00-a67ukykjrw7m7ixizm5jyaf5d5slvlzk/spack-build-out.txt
```

### Information on your system

```
siddiq90@login37> spack debug report
* **Spack:** 0.16.2-3949-8831cb2eed
* **Python:** 3.6.12
* **Platform:** cray-sles15-zen3
* **Concretizer:** original
```

This is a copy of my spack.yaml it is still in development. 
```
siddiq90@login37> cat spack.yaml 
spack:
  view: false
  config:
    install_missing_compilers: true
    install_tree:
      root: /global/common/software/spackecp/perlmutter/e4s-21.08/software
    build_stage: $tempdir/spack-stage
    misc_cache: $tempdir/misc_cache
  mirrors:
    perlmutter-e4s-21.08: /global/common/software/spackecp/mirrors/perlmutter-e4s-21.08
  definitions:
  - compilers:
    - nvhpc@21.7
  - cuda_specs:
      #- amrex@21.10 +openmp +pic +shared +cuda cuda_arch=80
      #- caliper@2.6.0 +cuda cuda_arch=80
      - hypre@2.22.0 +openmp +cuda cuda_arch=80

  - gcc_specs:
    - adios2@2.7.1 +dataman +png ~mpi
    - aml@0.1.0
    - amrex@21.08
    - arborx@1.0
    - argobots@1.1
    - ascent@0.7.1
    - bolt@2.0
    - darshan-runtime@3.3.1 scheduler=slurm ~mpi
    - darshan-util@3.3.1
    - papi@6.0.0.1

  - gcc: ['%gcc@10.3.0']
  specs:
  - $compilers
  - matrix:
    - [$cuda_specs]
    - ['%nvhpc']
  - matrix:
    - [$gcc_specs]
    - ['%gcc']
  compilers:
  - compiler:
      spec: gcc@10.3.0
      paths:
        cc: /opt/cray/pe/gcc/10.3.0/bin/gcc
        cxx: /opt/cray/pe/gcc/10.3.0/bin/g++
        f77: /opt/cray/pe/gcc/10.3.0/bin/gfortran
        fc: /opt/cray/pe/gcc/10.3.0/bin/gfortran
      operating_system: sles15
      modules:
      - PrgEnv-gnu/8.1.0
      - gcc/10.3.0
  - compiler:
      spec: nvhpc@21.7
      paths:
        cc: /global/common/software/spackecp/perlmutter/e4s-21.08/software/cray-sles15-zen3/gcc-10.3.0/nvhpc-21.7-sauysez5f7jr7oxutaa4i7opz4cpuq2y/Linux_x86_64/21.7/compilers/bin/nvc
        cxx: /global/common/software/spackecp/perlmutter/e4s-21.08/software/cray-sles15-zen3/gcc-10.3.0/nvhpc-21.7-sauysez5f7jr7oxutaa4i7opz4cpuq2y/Linux_x86_64/21.7/compilers/bin/nvc++
        f77: /global/common/software/spackecp/perlmutter/e4s-21.08/software/cray-sles15-zen3/gcc-10.3.0/nvhpc-21.7-sauysez5f7jr7oxutaa4i7opz4cpuq2y/Linux_x86_64/21.7/compilers/bin/nvfortran
        fc: /global/common/software/spackecp/perlmutter/e4s-21.08/software/cray-sles15-zen3/gcc-10.3.0/nvhpc-21.7-sauysez5f7jr7oxutaa4i7opz4cpuq2y/Linux_x86_64/21.7/compilers/bin/nvfortran
      flags: {}
      operating_system: sles15
      target: any
      environment: {}
      extra_rpaths: []
  packages:
    all:
      providers:
        compiler: [nvhpc@21.7, gcc@10.3.0]
        mpi: [mpich]
        fftw: [cray-fftw]
        blas: [cray-libsci]
        scalapack: [cray-libsci]
        pkgconfig: [pkg-config]
    amrex:
      variants: +openmp +pic +fortran
    bzip2:
      version: [1.0.6]
      buildable: false
      externals:
      - spec: bzip2@1.0.6
        prefix: /usr
    cray-libsci:
      buildable: false
      externals:
      - spec: cray-libsci@20.06.1
        modules:
        - cray-libsci/21.08.1.2
    cuda:
      buildable: false
      version: [11.0.2]
      externals:
      - spec: cuda@11.0.2
        prefix: /global/common/software/spackecp/perlmutter/e4s-21.08/software/cray-sles15-zen3/gcc-10.3.0/nvhpc-21.7-sauysez5f7jr7oxutaa4i7opz4cpuq2y/Linux_x86_64/21.7/cuda/11.0
    diffutils:
      version: [3.6]
      buildable: false
      externals:
      - spec: diffutils
        prefix: /usr
    cmake:
      buildable: false
      externals:
      - spec: cmake@3.20.5
        modules:
        - cmake/3.20.5
    cray-fftw:
      version: [3.3.8.8]
      buildable: false
      externals:
      - spec: cray-fftw@3.3.8.8
        modules:
        - cray-fftw/3.3.8.11
    hdf5:
      variants: +cxx +fortran
    gettext:
      version: [0.19.8.1]
      buildable: false
      externals:
      - spec: gettext
        prefix: /usr
    libfabric:
      buildable: false
      externals:
      - spec: libfabric@1.11.0
        modules:
        - libfabric/1.11.0.4.79
    mpich:
      buildable: false
      externals:
      - spec: mpich@3.4.1
        modules:
        - mpich/3.4.1
    m4:
      buildable: false
      externals:
      - spec: m4@1.4.18
        prefix: /usr
    openssl:
      buildable: false
      externals:
      - spec: openssl@1.1.1k
        prefix: /usr
    openssh:
      version: [7.9p1]
      buildable: false
      externals:
      - spec: openssh
        prefix: /usr
    openmpi:
      version: [4.0.5]
      variants: +cxx schedulers=slurm
    perl:
      version: [5.31.7]
    python:
      buildable: false
      externals:
      - spec: python@3.8.8
        modules:
        - python/3.9-anaconda-2021.05
    slurm:
      buildable: false
      version: [20-11-5-1]
      externals:
      - spec: slurm
        prefix: /usr
    tar:
      version: [1.3]
      buildable: false
      externals:
      - spec: tar@1.30
        prefix: /usr

```

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
Wasn't this solved by #26827 ?

**Comment by sethrj** [comment]
@shahzebsiddiqui This should have been fixed by  #25910. The spack version you're using is very old.

**Comment by shahzebsiddiqui** [comment]
okay i will try to cherry pick this PR and see if it fix the issue. I am building e4s/21.08 so this is based on spack branch https://github.com/spack/spack/tree/e4s-21.08

**Comment by shahzebsiddiqui** [comment]
> Wasn't this solved by #26827 ?

yes i had to use this PR to get started on perlmutter. This issue was specific to kokkos when trying to install on zen3

**Comment by alalazo** [comment]
See https://github.com/spack/spack/issues/26950#issuecomment-953178556
